File: killer_sudoku/image/validation.py
"""Stage 2 validation: flood-fill cages and produce a validated PuzzleSpec.

This module is the boundary between image-specific extraction (Stage 1) and the
newspaper-agnostic solver (Stage 3). All cage-layout consistency checks live
here so neither InpImage nor Grid needs to duplicate them.
"""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

from killer_sudoku.solver.grid import BRDR_MV, ProcessingError
from killer_sudoku.solver.puzzle_spec import PuzzleSpec, build_brdrs


def validate_cage_layout(
    cage_totals: npt.NDArray[np.intp],
    border_x: npt.NDArray[np.bool_],
    border_y: npt.NDArray[np.bool_],
) -> PuzzleSpec:
    """Flood-fill cage regions, validate each cage, and return a PuzzleSpec.

    Stage 2 of the pipeline: takes raw image-extraction output and produces the
    validated PuzzleSpec contract that the solver accepts.

    Three consistency checks are applied:
    - region_reassigned: a cell is reachable from two different cage heads,
      indicating an inconsistent border map.
    - invalid_cage: the declared cage total is outside the achievable range for
      the flood-filled cage size (too small or too large).
    - unassigned_region: at least one cell was not reached by any cage head,
      meaning the cage-total array is incomplete.

    Args:
        cage_totals: (9,9) array; non-zero at the top-left cell of each cage.
        border_x: (9,8) horizontal cage-wall flags from border detection.
        border_y: (8,9) vertical cage-wall flags from border detection.

    Returns:
        A fully-validated PuzzleSpec ready for Grid.set_up.

    Raises:
        ProcessingError: if regions overlap or a cell is left unassigned.
        ValueError: if a cage total is outside the achievable range for its size.
    """
    brdrs = build_brdrs(border_x, border_y)
    regions: npt.NDArray[np.intp] = np.zeros((9, 9), dtype=np.intp)

    def _flood(i: int, j: int, reg: int) -> int:
        """Recursively mark cell (i, j) as belonging to cage reg.

        Returns the count of cells newly marked in this call tree.
        Raises ProcessingError if a cell already assigned to a different cage
        is reached, indicating an inconsistent border map.
        """
        if regions[i][j] == 0:
            regions[i][j] = reg
            count = 1
            for b, mv in zip(brdrs[j][i], BRDR_MV, strict=False):
                if not b:
                    count += _flood(i + mv[1], j + mv[0], reg)
            return count
        if regions[i][j] != reg:
            raise ProcessingError("region reassigned", regions, brdrs)
        return 0

    reg = 0
    for i in range(9):
        for j in range(9):
            if cage_totals[i][j] != 0:
                reg += 1
                n = _flood(i, j, reg)
                lo = (n * (n + 1)) // 2
                hi = (n * (19 - n)) // 2
                if not (lo <= cage_totals[i][j] <= hi):
                    raise ValueError(
                        f"cagesize={n}, total={cage_totals[i][j]}: "
                        f"total must be in [{lo}, {hi}]"
                    )

    if (regions == 0).any():
        raise ProcessingError("unassigned region", regions, brdrs)

    return PuzzleSpec(
        regions=regions,
        cage_totals=cage_totals,
        border_x=border_x,
        border_y=border_y,
    )

File: web/src/image/validation.ts
/**
 * Union-find cage validation: Stage 2 of the image pipeline.
 *
 * Mirrors Python's `killer_sudoku.image.validation.validate_cage_layout`.
 * Takes raw border arrays from the clustering stage and produces a validated
 * PuzzleSpec, or throws if the cage layout is inconsistent.
 */

import { ProcessingError } from '../solver/errors.js';
import { buildBrdrs } from '../solver/puzzleSpec.js';
import type { PuzzleSpec } from '../solver/puzzleSpec.js';



/**
 * Build a string key for a cell to use as a Map key.
 * Using a string key avoids reference-equality pitfalls with arrays.
 */
function rowColKey(row: number, col: number): string {
  return `${row},${col}`;
}

/**
 * Union-find cage regions, validate each cage, and return a PuzzleSpec.
 *
 * Connected components are found by union-find directly on borderX/borderY,
 * avoiding the coordinate-convention pitfalls of the brdrs (9,9,4) expansion.
 *
 * Three consistency checks are applied:
 *   - region_reassigned: two cage heads map to the same connected component.
 *   - invalid_cage: the declared cage total is outside the achievable range for
 *     the connected-component size (too small or too large).
 *   - unassigned_region: at least one cell belongs to a component with no cage
 *     head, meaning the cage-total array is incomplete.
 *
 * @param cageTotals - (9, 9) array [col][row]; non-zero at the top-left of each cage.
 * @param borderX - (9, 8) horizontal cage-wall flags [col][rowGap].
 *   borderX[col][rowGap] = true means a wall between rows rowGap and rowGap+1.
 * @param borderY - (8, 9) vertical cage-wall flags [colGap][row].
 *   borderY[colGap][row] = true means a wall between cols colGap and colGap+1.
 * @returns A fully-validated PuzzleSpec ready for the solver.
 * @throws {ProcessingError} if cage heads clash or a cell is left unassigned.
 * @throws {Error} if a cage total is outside the achievable range for its size.
 */
export function validateCageLayout(
  cageTotals: number[][],
  borderX: boolean[][],
  borderY: boolean[][],
): PuzzleSpec {
  // Union-find: rmap[rowColKey] -> representative cell key.
  // members[repKey] -> set of cell keys in that component.
  const rmap = new Map<string, string>();
  const members = new Map<string, Set<string>>();

  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      const k = rowColKey(r, c);
      rmap.set(k, k);
      members.set(k, new Set([k]));
    }
  }

  function find(k: string): string {
    return rmap.get(k)!;
  }

  function union(ak: string, bk: string): void {
    const ra = find(ak);
    const rb = find(bk);
    if (ra === rb) return;
    // Always keep the lexicographically smaller key as representative.
    const [keep, drop] = ra < rb ? [ra, rb] : [rb, ra];
    for (const p of members.get(drop)!) {
      rmap.set(p, keep);
    }
    const keepSet = members.get(keep)!;
    for (const p of members.get(drop)!) {
      keepSet.add(p);
    }
    members.delete(drop);
  }

  // Merge cells across open horizontal borders (no wall between rows).
  // borderX[col][rowGap] = true means wall between rows rowGap and rowGap+1 in col.
  for (let col = 0; col < 9; col++) {
    for (let rowGap = 0; rowGap < 8; rowGap++) {
      if (!borderX[col]![rowGap]!) {
        union(rowColKey(rowGap, col), rowColKey(rowGap + 1, col));
      }
    }
  }

  // Merge cells across open vertical borders (no wall between columns).
  // borderY[colGap][row] = true means wall between colGap and colGap+1 in row.
  for (let colGap = 0; colGap < 8; colGap++) {
    for (let row = 0; row < 9; row++) {
      if (!borderY[colGap]![row]!) {
        union(rowColKey(row, colGap), rowColKey(row, colGap + 1));
      }
    }
  }

  const brdrs = buildBrdrs(borderX, borderY);

  // (9×9) regions array, [row][col], 1-based cage indices.
  const regions: number[][] = Array.from({ length: 9 }, () => new Array<number>(9).fill(0));
  let reg = 0;

  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      if (cageTotals[row]![col]! !== 0) {
        const repKey = find(rowColKey(row, col));
        const component = members.get(repKey)!;

        // Check no cell in this component has already been assigned.
        for (const k of component) {
          const [r, c] = k.split(',').map(Number) as [number, number];
          if (regions[r]![c]! !== 0) {
            throw new ProcessingError('region reassigned', regions, brdrs);
          }
        }

        reg += 1;
        const n = component.size;
        const lo = (n * (n + 1)) / 2;
        const hi = (n * (19 - n)) / 2;
        const total = cageTotals[row]![col]!;
        if (total < lo || total > hi) {
          throw new Error(
            `cagesize=${n}, total=${total} at row=${row + 1},col=${col + 1}: must be in [${lo}, ${hi}]`,
          );
        }

        for (const k of component) {
          const [r, c] = k.split(',').map(Number) as [number, number];
          regions[r]![c] = reg;
        }
      }
    }
  }

  // Check all cells have been assigned to a cage.
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      if (regions[r]![c]! === 0) {
        throw new ProcessingError('unassigned region', regions, brdrs);
      }
    }
  }

  return { regions, cageTotals, borderX, borderY };
}

/**
 * Clamp cage totals that are outside the achievable range for their cage size.
 *
 * Runs the same union-find as validateCageLayout to determine cage sizes, then
 * sets any head-cell total that is below lo or above hi to lo (the minimum).
 * Returns the repaired array and a list of human-readable warnings.
 *
 * Use this as a fallback when validateCageLayout throws an invalid_cage error:
 *   const { repaired, warnings } = repairCageTotals(totals, bx, by);
 *   spec = validateCageLayout(repaired, bx, by);
 */
export function repairCageTotals(
  cageTotals: number[][],
  borderX: boolean[][],
  borderY: boolean[][],
): { repaired: number[][]; warnings: string[] } {
  const rmap = new Map<string, string>();
  const members = new Map<string, Set<string>>();
  for (let r = 0; r < 9; r++) {
    for (let c = 0; c < 9; c++) {
      const k = rowColKey(r, c);
      rmap.set(k, k);
      members.set(k, new Set([k]));
    }
  }
  const find = (k: string) => rmap.get(k)!;
  const union = (ak: string, bk: string) => {
    const ra = find(ak); const rb = find(bk);
    if (ra === rb) return;
    const [keep, drop] = ra < rb ? [ra, rb] : [rb, ra];
    for (const p of members.get(drop)!) rmap.set(p, keep);
    const ks = members.get(keep)!;
    for (const p of members.get(drop)!) ks.add(p);
    members.delete(drop);
  };
  for (let col = 0; col < 9; col++)
    for (let rowGap = 0; rowGap < 8; rowGap++)
      if (!borderX[col]![rowGap]!) union(rowColKey(rowGap, col), rowColKey(rowGap + 1, col));
  for (let colGap = 0; colGap < 8; colGap++)
    for (let row = 0; row < 9; row++)
      if (!borderY[colGap]![row]!) union(rowColKey(row, colGap), rowColKey(row, colGap + 1));

  const repaired = cageTotals.map(row => [...row]);
  const warnings: string[] = [];
  for (let row = 0; row < 9; row++) {
    for (let col = 0; col < 9; col++) {
      const total = repaired[row]![col]!;
      if (total === 0) continue;
      const n = members.get(find(rowColKey(row, col)))!.size;
      const lo = (n * (n + 1)) / 2;
      const hi = (n * (19 - n)) / 2;
      if (total < lo || total > hi) {
        const clamped = total < lo ? lo : hi;
        repaired[row]![col] = clamped;
        warnings.push(`row=${row + 1},col=${col + 1}: read ${total}, clamped to ${clamped} (${n}-cell cage range [${lo}, ${hi}])`);
      }
    }
  }
  return { repaired, warnings };
}
